sklearn/cluster/k_means_.py

```diff
@@ -685,7 +685,9 @@ def _init_centroids(X, k, init, random_state=None, x_squared_norms=None,
         seeds = random_state.permutation(n_samples)[:k]
         centers = X[seeds]
     elif hasattr(init, '__array__'):
-        centers = init
+        # ensure that the centers have the same dtype as X
+        # this is a requirement of fused types of cython
+        centers = np.array(init, dtype=X.dtype)
```

Comment:
Ouch, that means a copy is already triggered before `_init_centroids` is called right?

Correct me if I'm wrong but in other words, if `X` is of dtype `np.float32`, there are 2 copies made of `init`.

1. One in `kmeans` which makes init of dtype `np.float64`
2. And then one in `_init_centroids` that converts it back to `np.float32`